torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -61,13 +61,29 @@ void checkSingleTensor(const std::vector<at::Tensor>& tensors) {
   if (!tensors[0].is_contiguous()) {
     throw std::runtime_error("input tensor has to be contiguous");
   }
+  if (tensors[0].is_sparse()) {
+    throw std::runtime_error("input tensor has to be dense");
+  }
   if (tensors[0].is_cuda() && !cudaAwareMpiCheck()) {
     throw std::runtime_error(
         "CUDA tensor detected and the MPI used doesn't "
         "have CUDA-aware MPI support");
   }
 }
 
+void checkSameSizeAndType(
+    const at::Tensor& tensor,
+    const std::vector<at::Tensor>& tensors) {
+  for (size_t i = 0; i < tensors.size(); ++i) {
+    if ((tensors[i].numel() != tensor.numel()) ||
+        (tensors[i].type() != tensor.type())) {
+      throw std::runtime_error("Tensors are not equal in size or data type");
+    }
+    std::vector<at::Tensor> temp{tensors[i]};
+    checkSingleTensor(temp);
```

Comment:
Can you change this function to take an `ArrayRef<at::Tensor>`? Right now you do an allocation of a single slot to hold `tensors[i]` at every iteration of this loop!

Reply:
I refactored